google-cloud-clients/google-cloud-storage/src/main/java/com/google/cloud/storage/BucketInfo.java

```diff
@@ -93,12 +97,242 @@ public BucketInfo apply(com.google.api.services.storage.model.Bucket pb) {
+                            }
+                          }));
+
+      return new LifecycleRule(lifecycleAction, conditionBuilder.build());
+    }
+
+    public static class LifecycleCondition implements Serializable {
+      private static final long serialVersionUID = -6482314338394768785L;
+      private Integer age;
+      private DateTime createdBefore;
+      private Integer numberOfNewerVersions;
+      private Boolean isLive;
+      private List<StorageClass> matchesStorageClass;
+
+      private LifecycleCondition(Builder builder) {
+        this.age = builder.age;
+        this.createdBefore = builder.createdBefore;
+        this.numberOfNewerVersions = builder.numberOfNewerVersions;
+        this.isLive = builder.isLive;
+        this.matchesStorageClass = builder.matchesStorageClass;
+      }
+
+      public Integer getAge() {
+        return age;
+      }
+
+      public DateTime getCreatedBefore() {
+        return createdBefore;
+      }
+
+      public Integer getNumberOfNewerVersions() {
+        return numberOfNewerVersions;
+      }
+
+      public Boolean getIsLive() {
+        return isLive;
+      }
+
+      public List<StorageClass> getMatchesStorageClass() {
+        return matchesStorageClass;
+      }
+
+      public static class Builder {
+        private Integer age;
+        private DateTime createdBefore;
+        private Integer numberOfNewerVersions;
+        private Boolean isLive;
+        private List<StorageClass> matchesStorageClass;
+
+        public Builder setAge(Integer age) {
+          this.age = age;
+          return this;
+        }
+
+        public Builder setCreatedBefore(DateTime createdBefore) {
+          this.createdBefore = createdBefore;
+          return this;
+        }
+
+        public Builder setNumberOfNewerVersions(Integer numberOfNewerVersions) {
+          this.numberOfNewerVersions = numberOfNewerVersions;
+          return this;
+        }
+
+        public Builder setIsLive(Boolean live) {
+          this.isLive = live;
+          return this;
+        }
+
+        public Builder setMatchesStorageClass(List<StorageClass> matchesStorageClass) {
+          this.matchesStorageClass = matchesStorageClass;
+          return this;
+        }
+
+        public LifecycleCondition build() {
+          return new LifecycleCondition(this);
+        }
+      }
+    }
+
+    public abstract static class LifecycleAction implements Serializable {
+      private static final long serialVersionUID = 5801228724709173284L;
+
+      abstract String getActionType();
+    }
+
+    public static class DeleteLifecycleAction extends LifecycleAction {
+      public static final String TYPE = "Delete";
+      private static final long serialVersionUID = -2050986302222644873L;
+
+      @Override
+      String getActionType() {
+        return TYPE;
+      }
+    }
+
+    public static class SetStorageClassLifecycleAction extends LifecycleAction {
```

Comment:
This should follow a builder pattern in case of changes made to this action. 

I don't expect changes to come to actions, but based on our offline discussion it would help here.